### PR TITLE
Switch to unclog for the changelog.

### PR DESCRIPTION
**Pull request by SpicyLemon**
## Description

This PR will switch us to using [unclog](https://github.com/informalsystems/unclog) for our unreleased changelog information.

After this PR is merged, you should no longer add new entries to the `CHANGELOG.md` file, but instead create new unreleased entry files using `unclog`.

Example:

```console
$ unclog add -p 2112 -i switch-to-unclog -s improvements -m 'Switch to `unclog` for unreleased changelog entries'
23:08:32 [INFO] Attempting to load configuration file from: .changelog/config.toml
23:08:32 [INFO] Deduced GitHub project!
23:08:32 [INFO] Loading change template from: .changelog/change-template.md
23:08:32 [INFO] Created directory: .changelog/unreleased/improvements
23:08:32 [INFO] Wrote entry to: .changelog/unreleased/improvements/2112-switch-to-unclog.md
```

If your change involves dependency changes, you should utilize the `.changelog/get-dep-changes.sh` script (instead of `unclog`) to generate the dependency entry file (more info below). This should be in addition to one or more entries specific to your change.

Example:
```console
$ ./.changelog/get-dep-changes.sh -p 2112 -i bump-thing-to-whatever
Dependency changelog entry created: .changelog/unreleased/dependencies/2112-bump-thing-to-whatever.md
```

To get a list of valid `--section` values, run `make get-valid-sections` (or `.changelog/get-valid-sections.sh`).

Example:

```console
$ .changelog/get-valid-sections.sh
features
improvements
bug-fixes
deprecated
client-breaking
api-breaking
state-machine-breaking
dependencies
```

A new linter script has been added (`.changelog/lint-unreleased.sh`) that fails if there's an unexpected section directory in `unreleased` or if there are any files under that dir with a space in the filename. The `make lint` target now also runs this linter, as well as the GitHub action.

We'll still use the `CHANGELOG.md` as a source of truth though; `unclog` is only being used to manage unreleased dependencies and generate _new_ release content. We won't be using it to create the entirety of the changelog. As such, when updating the `CHANGELOG.md` file on `main`, we'll also be deleting the associated `unreleased` entry files instead of moving them to a version directory.

---

For more thorough dependency change entries, I've created a script to analyze changes being made to `go.mod` and generate all the appropriate dependency changelog entries: `.changelog/get-dep-changes.sh`. This script is designed for you to use. It will only identify dependency changes involving the `go.mod` file. The dependabot changelog action utilizes a new `scripts/dependabot-changelog.sh` which calls `get-dep-changes.sh` but if there are no changes to `go.mod`, it then creates an appropriate changelog entry based off the title of the PR.

Before this PR, if dependabot bumped one library that caused others to be bumped (e.g. when doing `go mod tidy`), the changelog only had an entry for the library that initiated the PR. Now, though, we'll have a more detailed listing of the changes.

Further, the format of the entries have changed a little bit. Entries now have the format `<library> <action> <version> [(from <old version>)]`. The `<action>` will be one of `bumped to`, `added at`, or `removed at`. The `(from <old version>)` part is only included for version bumps. If the library was removed, the `<version>` is what we were using right before we removed the dependency. Replace lines are now considered for these entries too. If a replace line is involved, the `<version>` and `<old version>` strings can also contain the replacing library.

Whenever your PR involves a `go.mod` change, you should run `scripts/get-dep-changes.sh` to generate the changelog entries (instead of using `unclog` for the dependency changes).

If your change contains a library bump, but that bump isn't the purpose of the change, you should create both a `dependencies` entry (using this script) and an entry more specific to your change (using `unclog`).

I tested the github action in this PR by doing the following:

1. I added the "dependencies" label to this PR so that the changelog action would run.
2. I let the action run with a bunch of complex changes in `go.mod`: [d781915](https://github.com/provenance-io/provenance/pull/2112/commits/d78191569314aa4ee8b0edb58ab7dd31caafc9e8).
3. I verified that it created a new file in the correct directory with the expected content based on the go.mod changes: [3035b86](https://github.com/provenance-io/provenance/pull/2112/commits/3035b860a4946e68b5ef3aa6abb4cbc70ad35ae0).
4. I changed this PR's title to something closer to what Dependabot does: `"Bump some/thing from v1.2.3 to v1.2.4"`.
5. I changed `go.mod` back to what `main` has and pushed that up: [b91decd](https://github.com/provenance-io/provenance/pull/2112/commits/b91decdb3e65db2b5b7a2f3026920cc32891bbba).
6. I verified that it created a new file in the correct directory with the expected content based on the PR title: [b420cb7](https://github.com/provenance-io/provenance/pull/2112/commits/b420cb78fc2322b260c3430566d47087e301637e).
7. I deleted that changelog file and changed this PR's title back.

---

I've also created a new script that will make the updates that are needed to mark a new release: `.changelog/prep-release.sh`. This script will create/update the `RELEASE_NOTES.md` file, add the version to the `CHANGELOG.md` file, and move the entry files out of `.changelog/unreleased/`.

This script handles the following cases (at the very least):
1. Creating a new rc1.
2. Creating subsequent release candidates.
3. Creating a full release for a version that had release candidates.
4. Creating a full release for a version that did **NOT** have a release candidate.

I've updated the `CONTRIBUTING.md` release procedure documentation to include use of this script.

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/provenance-io/provenance/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Wrote unit and integration [tests](https://github.com/provenance-io/provenance/blob/main/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

## Summary by CodeRabbit

- **New Features**
	- Introduced a structured changelog management system with templates and automation scripts for better documentation.
	- Added automation scripts for generating changelog entries from Dependabot updates and validating changelog formatting.

- **Documentation**
	- Enhanced the `CHANGELOG.md` and `.changelog/README.md` to clarify the changelog process and entry guidelines.
	- Updated `CONTRIBUTING.md` for clarity and consistency in instructions.
	- Revised the pull request template to emphasize changelog entry requirements.

- **Bug Fixes**
	- Improved error handling and validation in automation scripts to enhance reliability during changelog updates.

### CODE REVIEW

File: .changelog/dependabot-changelog.sh [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+    printf 'No --pr <num> provided.\n'
+    stop_early='YES'
+fi
+
+if [[ -z "$title" ]]; then
+    printf 'No --title <title> provided.\n'
+    stop_early='YES'
+fi
+
+[[ -n "$stop_earl" ]] && exit 1
```
[the remark this answers is not in the document]
Comment (SpicyLemon) [reply .changelog/dependabot-changelog.sh]: Fixed! Thanks.

File: .changelog/dependabot-changelog.sh [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+ec=$?
+[[ -n "$verbose" ]] && printf 'Exit code from get-dep-changes.sh: %d\n' "$ec"
+
+# That script exits with 0 when there are go.mod changes and the new file was created.
+# If there were go.mod changes, we're all done here.
+# I don't think I've ever seen a dependabot PR that bumps both a go module and something else.
+if [[ "$ec" -eq '0' ]]; then
+    [[ -n "$verbose" ]] && printf 'Changes identified through go.mod. Done.\n'
+    exit 0
+fi
```
[the remark this answers is not in the document]
Comment (SpicyLemon) [reply .changelog/dependabot-changelog.sh]: Fixed. Thanks!

File: .changelog/get-dep-changes.sh [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+        [[ -n "$verbose" ]] && printf 'Making dir (if it does not exist yet): %s\n' "$out_dir"
+        mkdir -p "$out_dir" || failed="YES"
+    fi
+    if [[ -z "$failed" ]]; then
+        [[ -n "$verbose" ]] && printf 'Copying final file from %s to %s\n' "$final" "$out_fn"
+        if cp "$final" "$out_fn"; then
+            copied='YES'
+            printf 'Dependency changelog entry created: %s\n' "$out_fn"
+        fi
+    fi
```
[the remark this answers is not in the document]
Comment (SpicyLemon) [reply .changelog/get-dep-changes.sh:374]: The shorter "Making dir" is just fine.

**Comment by iramiller** [comment]
The `.github/PULL_REQUEST_TEMPLATE.md` file should also be updated with appropriate checkboxes for these new `unclog` steps that are required (either a `PR` or a `issue-no` must be given with `unclog` on the PR)

**Comment by SpicyLemon** [comment]
> The `.github/PULL_REQUEST_TEMPLATE.md` file should also be updated with appropriate checkboxes for these new `unclog` steps that are required (either a `PR` or a `issue-no` must be given with `unclog` on the PR)

Good call. I've tweaked it.
Was: ``Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md` ``
Now: ``Added relevant changelog entries under `.changelog/unreleased` (see [Adding Changes](.changelog/README.md#adding-changes)).``

**Review by iramiller — APPROVED** [review]
Some seriously complex scripting in here... seems like it should work out ok but the proof will come with use.

**Comment by SpicyLemon** [comment]
> Some seriously complex scripting in here... seems like it should work out ok but the proof will come with use.

Yeah. Luckily, both `dependabot-changelog.sh` and `get-dep-changes.sh` will get solid tests right away with the existing dependabot PRs. In this branch, I did some pretty extensive testing of all of that stuff. I even added the `dependencies` label to this PR and made some crazy changes to `go.mod`, and also played with the PR title (without `go.mod` changes) to test the updated changelog GitHub action.

And `lint-unreleased.sh` makes use of `get-valid-sections.sh` and is now run in the lint GitHub action. I tested that too, by adding all the valid sections, then a few invalid ones. And also added some entries with invalid filenames.

The `prep-release.sh` won't get any action until our next release though, but I did some pretty extensive testing of it in this branch, so I'm pretty confident it'll work just fine for at least our next few releases. If not, at least I detailed all the stuff that needs to happen, and doing it manually wouldn't be too horrible.

About the only thing that concerns me a little is the post-release PR to update the changelog on `main`. I'm hoping git makes it easy to delete the stuff from `unreleased` that was just released without touching entries that weren't in the release, but we'll have to wait and see.

**Review by nullpointer0x00 — APPROVED** [review]

**Review by Taztingo — APPROVED** [review]

**Pull request merged**